Automattic/jetpack
# Google Analytics: Add remaining enhanced eCommerce trackers

Fixes #8181 

#### Changes proposed in this Pull Request:

* Adds support for the following Enhanced eCommerce tracking events:

* Remove item from cart
* Track product impressions
* Track product listing clicks
* Track product detail views
* Track checkout started

#### Testing instructions:

* See also https://github.com/Automattic/jetpack/pull/8182
* Sign up for a Google Analytics tracking code for your test site
* Install, activate and connect Jetpack
* Install and activate WooCommerce
* Create a product with no SKU and at least two categories
* Create another product with a SKU and at least two categories
* Replace modules/google-analytics with the files from this branch
* Install and activate the Jetpack Google Analytics Helper plugin ( https://github.com/Automattic/jetpack-google-analytics-helper )

#### Remove From Cart

* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: UA-12345678-1 (use a real tracking/property ID, not this one)
Anonymize IP Addresses: unchecked
Track Purchase Transactions: unchecked
Track Add to Cart Events: unchecked
Enable Enhanced eCommerce: CHECKED
Track Remove from Cart: CHECKED
Track Product Impressions on Listing Pages: unchecked
Track Product Clicks on Listing Pages: unchecked
Track Product Detail Views: unchecked
Track Checkout Process Initiated: unchecked
```

* In your Google Analytics Dashboard ( https://analytics.google.com/ ), go to Admin for your site (it looks like a gear in the lower left corner)
* Under View select Ecommerce Settings
* Enable Ecommerce AND Enhanced eCommerce if you haven't already.

* Add the SKU-less product and the SKU'd product to the cart
* Navigate to cart view
* View the source for the page
* Search for "ec:" - ensure javascript is enqueued for both products
* Open the JavaScript dev console
* Click on the remove for each product in turn
* Ensure no JavaScript errors appear in the console
* Wait a few minutes
* In your Google Analytics Dashboard, go to Real-time and then Events
* Make sure you see the remove from cart events for the SKU-less and the SKU'd product (you might need to click on the UX category to see them)

<img width="1299" alt="screen shot 2017-11-20 at 5 11 23 pm" src="https://user-images.githubusercontent.com/1595739/33049628-e2f30732-ce15-11e7-9832-7b134c547863.png">

#### Product Impressions

* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: UA-12345678-1 (use a real tracking/property ID, not this one)
Anonymize IP Addresses: unchecked
Track Purchase Transactions: unchecked
Track Add to Cart Events: unchecked
Enable Enhanced eCommerce: CHECKED
Track Remove from Cart: unchecked
Track Product Impressions on Listing Pages: CHECKED
Track Product Clicks on Listing Pages: unchecked
Track Product Detail Views: unchecked
Track Checkout Process Initiated: unchecked
```

* Open the JavaScript dev console if not already open
* Navigate to shop view (i.e. products list)
* View the source for the page
* Search for "ec:addImpression" - make sure that ec: command is enqueued for both products
* Ensure no JavaScript errors appear in the console
* Wait a few minutes
* In your Google Analytics Dashboard, go to Conversions and then eCommerce and then Product List Performance
* Make sure you see the entries for the SKU-less and the SKU'd product

<img width="1312" alt="screen shot 2017-11-20 at 5 17 07 pm" src="https://user-images.githubusercontent.com/1595739/33049821-aceb7cb8-ce16-11e7-88be-973b58b685ef.png">

#### Product Clicks

* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: UA-12345678-1 (use a real tracking/property ID, not this one)
Anonymize IP Addresses: unchecked
Track Purchase Transactions: unchecked
Track Add to Cart Events: unchecked
Enable Enhanced eCommerce: CHECKED
Track Remove from Cart: unchecked
Track Product Impressions on Listing Pages: unchecked
Track Product Clicks on Listing Pages: CHECKED
Track Product Detail Views: unchecked
Track Checkout Process Initiated: unchecked
```

* Open the JavaScript dev console if not already open
* Navigate to shop view (i.e. products list)
* View the source for the page
* Search for "ec:" - make sure that a click handler is in the page source for each product, e.g.

```
$( '.products .post-1390 a' ).click( function() {
```

* Ensure no JavaScript errors appear in the console
* Click on the PICTURE for each product (NOT the Add to Cart button) (and then go back and repeat for the next product)
* Wait a few minutes
* In your Google Analytics Dashboard, go to Conversions and then eCommerce and then Product List Performance
* Make sure you see the entries for the SKU-less and the SKU'd product

#### Product Detail View

* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: UA-12345678-1 (use a real tracking/property ID, not this one)
Anonymize IP Addresses: unchecked
Track Purchase Transactions: unchecked
Track Add to Cart Events: unchecked
Enable Enhanced eCommerce: CHECKED
Track Remove from Cart: unchecked
Track Product Impressions on Listing Pages: unchecked
Track Product Clicks on Listing Pages: unchecked
Track Product Detail Views: CHECKED
Track Checkout Process Initiated: unchecked
```

* Open the JavaScript dev console if not already open
* Navigate to a single product view
* View the source for the page
* Search for 'ec:setAction', 'detail' - make sure that an addProduct preceeds it for the product
* Ensure no JavaScript errors appear in the console
* Wait a few minutes
* In your Google Analytics Dashboard, go to Conversions and then eCommerce and then Product List Performance
* Make sure you see the entries for the SKU-less and the SKU'd product

#### Checkout Initiated

* Use the helper plugin (wp-admin > Tools > Jetpack Google Analytics Helper) to set the following:

```
Google Analytics Tracking / Property ID: UA-12345678-1 (use a real tracking/property ID, not this one)
Anonymize IP Addresses: unchecked
Track Purchase Transactions: unchecked
Track Add to Cart Events: unchecked
Enable Enhanced eCommerce: CHECKED
Track Remove from Cart: unchecked
Track Product Impressions on Listing Pages: unchecked
Track Product Clicks on Listing Pages: unchecked
Track Product Detail Views: unchecked
Track Checkout Process Initiated: CHECKED
```

* Open the JavaScript dev console if not already open
* Navigate to a each product view and add the product to your cart
* Proceed to checkout
* View the source for the page
* Search for 'ec:setAction', 'checkout' - make sure that an addProduct preceeds it for each product, e.g.:

```
ga( 'ec:addProduct', {"id":"12345678-0123","name":"Has SKU Two Cat","category":"Albums\/T-shirts","price":"123","quantity":1} );
ga( 'ec:addProduct', {"id":"#1389","name":"No SKU Two Cat","category":"Albums\/Singles","price":"123","quantity":1} );
ga( 'ec:setAction','checkout' );
```

* Ensure no JavaScript errors appear in the console
* Wait a few minutes
* In your Google Analytics Dashboard, go to ??? (I couldn't find where this ends up - do you know @justinshreve  ?)

* Lastly, disable WooCommerce
* Make sure with a tracker ID enabled, and all settings INCLUDING enhanced ecommerce enabled, that you are able to navigate the site without PHP errors
* Make sure with a tracker ID enabled, and all settings EXCEPT enhanced ecommerce enabled, that you are able to navigate the site without PHP errors
